script.js

```diff
@@ -0,0 +1,220 @@
+
+  const img = document.createElement("img");
+  img.src = recipe.image;
+  img.alt = recipe.title;
+
+  const title = document.createElement("h3");
+  title.textContent = recipe.title;
+
+  const link = document.createElement("a");
+  link.href = "#";
+  link.textContent = "View Recipe";
+  link.addEventListener("click", async (e) => {
+    e.preventDefault();
+    await showRecipeDetails(recipe.id);
+  });
+
+  div.append(img, title, link);
+  return div;
+}
+
+function setLoading(loading, message = "Searching…") {
+  if (loading) {
+    resultsEl.innerHTML = `<p>${message}</p>`;
+    searchBtn.disabled = true;
+    randomBtn.disabled = true;
+    searchBtn.style.opacity = 0.7;
+    randomBtn.style.opacity = 0.7;
+  } else {
+    searchBtn.disabled = false;
+    randomBtn.disabled = false;
+    searchBtn.style.opacity = "";
+    randomBtn.style.opacity = "";
+  }
+}
+
+async function searchRecipes() {
+  const ingredient = (inputEl?.value || "").trim();
+  if (!ingredient) {
+    resultsEl.innerHTML = "<p>Please enter an ingredient.</p>";
+    return;
+  }
+
+  const type = mealTypeEl.value;
+  const cuisine = countryEl.value;
+  const diet = dietEl.value;
+  const sort = sortEl.value;
+
+  try {
+    setLoading(true, "Searching…");
+
+    const qs = new URLSearchParams({
+      query: ingredient,
+      number: "9",
+      addRecipeInformation: "true",
+      apiKey
+    });
+
+    if (type) qs.set("type", type);
+    if (cuisine) qs.set("cuisine", cuisine);
+    if (sort) qs.set("sort", sort);
+
+    const dietParams = buildDietParams(diet);
+    dietParams.forEach((v, k) => qs.set(k, v));
+
+    const url = `https://api.spoonacular.com/recipes/complexSearch?${qs.toString()}`;
+    const response = await fetch(url);
+    if (!response.ok) {
+      const msg = response.status === 402 || response.status === 401
+        ? "API key error or quota reached. Please check your Spoonacular API key."
+        : `Request failed (HTTP ${response.status}).`;
+      throw new Error(msg);
+    }
+
+    const data = await response.json();
+    const items = data.results || [];
+
+    resultsEl.innerHTML = items.length
+      ? ""
+      : "<p>No recipes found. Try another ingredient or adjust filters.</p>";
+
+    items.forEach((r) => resultsEl.appendChild(card(r)));
+  } catch (err) {
+    console.error(err);
+    resultsEl.innerHTML = `<p>${err.message || "Something went wrong. Please try again."}</p>`;
+  } finally {
+    setLoading(false);
+  }
+}
+
+async function randomRecipe() {
+  const type = mealTypeEl.value;
+  const cuisine = countryEl.value;
+  const diet = dietEl.value;
+
+  try {
+    setLoading(true, "Picking a random recipe…");
+
+    const tags = [];
+    if (type) tags.push(type);
+    if (cuisine) tags.push(cuisine);
+    if (diet) tags.push(diet); // Spoonacular accepts "dairy free" tag too
+
+    const qs = new URLSearchParams({ number: "1", apiKey });
+    if (tags.length) qs.set("tags", tags.join(","));
```

Comment:
Nice use of URLSearchParams! we haven't talked about this concept yet but it is a very lcean way of structuring the code. also to set the query string with the tags like that. ⭐